server.go

```diff
@@ -0,0 +1,264 @@
+type SecurityEntry struct {
+	Version   SNMPVersion // SNMP version to use (V2c only)
+	Community string      // Community
+}
+
+func (a *SecurityEntry) validate() error {
+	if a.Version != V2c {
+		return &ArgumentError{
+			Value:   a.Version,
+			Message: "Unsupported SNMP Version",
+		}
+	}
+	return nil
+}
+
+func (a *SecurityEntry) String() string {
+	return escape(a)
+}
+
+// TrapListener defines method that need to be implemented by Trap listeners.
+// If OnTRAP panics, the server (caller of OnTRAP) assumes that affect of the panic
+// is temporary and recovers by the panic and logs trace to the error log.
+type TrapListener interface {
+	OnTRAP(trap *TrapRequest)
+}
+
+// TrapRequest is representing trap request that is send from the network element.
+type TrapRequest struct {
+	// The received PDU
+	Pdu Pdu
+
+	// The source address of trap
+	Source net.Addr
+
+	// Error is an optional field used to indicate
+	// errors which may occur during the decoding
+	// of the received packet
+	Error error
+}
+
+// A TrapServer defines parameters for running of TRAP daemon that listens for incoming
+// trap messages.
+type TrapServer struct {
+	args      *ServerArguments
+	mp        messageProcessing
+	secs      *securityMap
+	transport transport
+	serving   bool
+
+	// Trap Listener
+	Listener TrapListener
+
+	// Error Logger which will be used for logging of default errors
+	ErrorLog *log.Logger
+}
+
+func (s *TrapServer) AddSecurity(entry *SecurityEntry) error {
+	if err := entry.validate(); err != nil {
+		return err
+	}
+	s.secs.Set(newSecurityFromEntry(entry))
+	return nil
+}
+
+func (s *TrapServer) DeleteSecurity(entry *SecurityEntry) {
+	s.secs.Delete(newSecurityFromEntry(entry))
+}
+
+func (s *TrapServer) Serve(listener TrapListener) error {
+	s.Listener = listener
+	s.serving = true
+	size := s.args.MessageMaxSize
+	if size < recvBufferSize {
+		size = recvBufferSize
+	}
+
+	for {
+		conn, err := s.transport.Listen()
+		if !s.serving {
+			return nil
+		}
+		if err != nil {
+			if e, ok := err.(net.Error); ok && e.Temporary() {
+				continue
+			}
+			return err
+		}
+
+		go func(conn interface{}) {
+			defer s.transport.Close(conn)
+			buf := make([]byte, size)
+			for {
+				_, src, msg, err := s.transport.Read(conn, buf)
+				if _, ok := err.(net.Error); ok {
+					if s.serving {
+						s.logf("trap: failed to read packet: %v", err)
+					}
+					return
+				}
+
+				go s.handle(conn, msg, src, err)
+			}
+		}(conn)
+	}
+}
+
+func (s *TrapServer) Close() error {
+	s.serving = false
+	return s.transport.Close(nil)
+}
+
+// handle a newly received trap
+func (s *TrapServer) handle(conn interface{}, msg message, src net.Addr, err error) {
+	defer func() {
+		if err := recover(); err != nil {
+			const size = 64 << 10
+			logBuf := make([]byte, size)
+			logBuf = logBuf[:runtime.Stack(logBuf, false)]
+			s.logf("trap: panic while receiving %v: %v\n%s", src, err, logBuf)
+
+		}
+	}()
+
+	l := s.Listener
```

Comment:
Listener can be changed at any time, because it is public member.
Fixed ed35def765edc79fa1ae8b6daf41c72941620027